### PR TITLE
Add :enabled property to syntax checkers

### PR DESCRIPTION
**Pull request by Simplify**
Closes GH-1089

### CODE REVIEW

File: flycheck.el [hunk 1]
```diff
@@ -295,6 +295,17 @@ respectively."
   :safe #'flycheck-symbol-list-p)
 (make-variable-buffer-local 'flycheck-disabled-checkers)
 
+(defcustom flycheck-enabled-checkers nil
```
Comment (lunaryorn) [inline flycheck.el]: I don't think that this should be a customizable variable, should it?  To me it's internal state only.  I'd turn it into a `defvar-local` and move it down beneath the definition of `flycheck-may-enable-checker`.

File: flycheck.el [hunk 2]
```diff
@@ -4408,6 +4459,13 @@ default `:verify' function of command checkers."
                                (and verify-fn
                                     (funcall verify-fn checker)))))))
 
+  (let ((enabled-fn (plist-get properties :enabled)))
+    (setq properties
+          (plist-put properties :enabled
+                     (lambda (checker)
+                       (and enabled-fn
+                            (funcall enabled-fn checker))))))
```
Comment (lunaryorn) [inline flycheck.el]: I don't think that this is needed is it? After all, we're not adding to `:enabled` here, do we?

File: test/flycheck-test.el [hunk 3]
```diff
@@ -3315,20 +3322,28 @@ Why not:
...
     (flycheck-ert-should-syntax-check
      "language/javascript/warnings.js" flycheck-test-javascript-modes
-     '(3 2 warning "Use the function form of \"use strict\"." :id "strict"
+     '(3 2 warning "Use the function form of 'use strict'." :id "strict"
          :checker javascript-eslint)
-     '(4 9 warning "\"foo\" is defined but never used" :id "no-unused-vars"
+     '(4 9 warning "'foo' is defined but never used." :id "no-unused-vars"
+         :checker javascript-eslint))))
+
+(flycheck-ert-def-checker-test javascript-eslint javascript warning-without-config
```
Comment (lunaryorn) [inline test/flycheck-test.el]: I thought it wouldn't work without configuration?  I'm confused 😕

File: flycheck.el [hunk 4]
```diff
@@ -7439,6 +7500,17 @@ See URL `https://github.com/eslint/eslint'."
...
+             (let* ((command (concat (flycheck-find-checker-executable checker)
+                                     (if flycheck-eslintrc
+                                         (concat " --config " flycheck-eslintrc)
+                                       "")
+                                     " --print-config "
+                                     (flycheck-compute-working-directory checker)
+                                     "."))
+                    (exitcode (call-process-shell-command command nil nil nil)))
+               (cond ((eq exitcode 0) nil)
+                     (t "Configuration not found"))))
```
Comment (lunaryorn) [inline flycheck.el]: Thanks, that's nice, but would you mind to remove the changes from the eslint checker again?  Let's keep this PR limited to `:enabled`, and make second one for the change to eslint?  That's just easier for me to follow, and I believe it's better to keep pull requests confined to just one change.

File: flycheck.el [hunk 5]
```diff
@@ -1762,7 +1812,8 @@ nil otherwise."
     (and (flycheck-valid-checker-p checker)
          (not (flycheck-disabled-checker-p checker))
          (flycheck-checker-supports-major-mode-p checker major-mode)
-         (funcall predicate))))
+         (funcall predicate)
+         (flycheck-may-enable-checker checker))))
```
Comment (lunaryorn) [inline flycheck.el]: I think we should move this before the `predicate` check, and we can merge it with the disable check above.

File: flycheck.el [hunk 6]
```diff
@@ -1753,6 +1789,20 @@ CHECKER supports buffers in the given major MODE.
 Return non-nil if CHECKER supports MODE and nil otherwise."
   (memq mode (flycheck-checker-get checker 'modes)))
 
+(defun flycheck-may-enable-checker (checker)
+  "Whether a generic CHECKER may be enabled for current buffer.
+
+Return non-nil if CHECKER may be used for the current buffer, and
+nil otherwise."
+  (if (member checker flycheck-enabled-checkers)
```
Comment (lunaryorn) [inline flycheck.el]: Use `memq`, as we're dealing with symbols.  Also, try not to use `if`.  This can be rewritten as an `and` expression that's easier to follow, imho:

```elisp
(let ((enabled (flycheck-checker-get checker 'enabled))
      (shall-enable (and (not (flycheck-disabled-checker-p checker))
                         (or (memq checker flycheck-enabled-checkers)
                             (null enabled)
                             (funcall enabled checker)))))
  (if shall-enable
      (cl-pushnew checker flycheck-enabled-checkers)
    (cl-pushnew checker flycheck-disabled-checkers))
  shall-enable)
```

What do you think?

File: doc/languages.rst [hunk 7]
```diff
@@ -549,6 +549,10 @@ to view the docstring of the syntax checker.  Likewise, you may use
 
       Check syntax and lint with `ESLint <http://eslint.org/>`_.
 
+      .. note::
+
+         This syntax checker requires presence of a configuration file.
```
Comment (cpitclaudel) [inline doc/languages.rst]: Should this be "requires *the* presence"?

File: flycheck.el [hunk 8]
```diff
@@ -1691,6 +1723,9 @@ Signal an error, if any property has an invalid value."
     (unless (or (null verify) (functionp verify))
       (error ":verify %S of syntax checker %S is not a function"
              symbol verify))
+    (unless (or (null enabled) (functionp enabled))
+      (error ":enabled %S of syntax checker %S is not a function"
+             symbol enabled))
```
Comment (cpitclaudel) [inline flycheck.el]: Are the two arguments flipped?

File: test/flycheck-test.el (hunk 3 above)
Comment (Simplify) [inline test/flycheck-test.el]: Better name is `with-automatic-config-selection`. Test is more to see if errors are correct if you provide config or not.

File: flycheck.el (hunk 4 above)
Comment (cpitclaudel) [inline flycheck.el]: One small addition: this code isn't safe wrt spaces in working directory or file name.  I think this code might be able to use `flycheck-substitute-argument`, and if not, `shell-quote-argument`

**Comment by Simplify** [comment]
@cpitclaudel Thanks, I'll fix it with `shell-quote-argument` in PR for ESLint.
Not sure what do you mean with flipped arguments.

I have done other suggestions/requests.

**Comment by cpitclaudel** [comment]
I meant that 

```
(error ":enabled %S of syntax checker %S is not a function" symbol enabled)
```

looked like it should be 

```
(error ":enabled %S of syntax checker %S is not a function" enabled symbol)
```

instead.  Maybe I'm confused?

**Comment by Simplify** [comment]
I that case it's wrong everywhere in that function, `:start`, `:interrupt`, `:print-doc`,... I simply copied `:verify` part and renamed variables...

**Comment by lunaryorn** [comment]
@Simplify  `:verify` and `:predicate` extend the given functions with default ones.  For instance `flycheck-define-command-checker` actually defines a *new* `:predicate` that uses the given `:predicate` function *and* `flycheck-find-checker-executable`.

For `:enabled` we do not need that as far as I can see.

**Comment by lunaryorn** [comment]
@Simplify Thank you for removing the eslint change, will take a new look at this soon.

**Comment by lunaryorn** [comment]
@Simplify Oh dear, it is :flushed: Thanks for making me aware, @cpitclaudel 

I pushed a fix 😌

**Comment by cpitclaudel** [comment]
Neat :) Though I'm not seeing the fix in the repo.

**Comment by lunaryorn** [comment]
Darn, I forgot to push 😳 Will do tomorrow

**Comment by lunaryorn** [comment]
So, finally pushed. Sorry.

**Comment by lunaryorn** [comment]
@Simplify I'll give this another review tomorrow evening.

File: doc/languages.rst [hunk 9]
```diff
@@ -549,6 +549,10 @@ to view the docstring of the syntax checker.  Likewise, you may use
 
       Check syntax and lint with `ESLint <http://eslint.org/>`_.
 
+      .. note::
+
+         This syntax checker requires the presence of a configuration file.
+
```
Comment (lunaryorn) [inline doc/languages.rst]: I guess that's for the other pull request as well, right?  Looks as if it's related to the eslint change.

File: flycheck.el [hunk 10]
```diff
@@ -1610,6 +1610,26 @@ are mandatory.
...
 
+`:enabled FUNCTION'
+     A function to determine whether to use the syntax checker in
+     the current buffer.
+
+     This function behaves as `:predicate', except that it's only
+     called the first time a syntax checker is to be used in a buffer.
+
+     If the function returns a non-nil value the checker is put in a
+     whitelist to prevent further invocations of `:enabled', and if it
```
Comment (lunaryorn) [inline flycheck.el]: Mention the variable name, i.e. "…is whitelisted in \`flycheck-enabled-checkers' to prevent…".  Also, "and if…" makes for a long sentence.  I'd start a new one, i.e "Otherwise it is disabled via \`flycheck-disabled-checkers' to prevent…"

File: flycheck.el [hunk 11]
```diff
@@ -1610,6 +1610,26 @@ are mandatory.
...
+     returns nil it's put into flycheck-disabled-checkers to prevent
```
Comment (lunaryorn) [inline flycheck.el]: Please quote variable names, i.e. "`flycheck-disabled-checkers'"

File: test/flycheck-test.el [hunk 12]
```diff
@@ -3315,20 +3322,20 @@ Why not:
...
   (let ((flycheck-eslintrc "eslint.json")
         (flycheck-disabled-checkers '(javascript-jshint javascript-jscs)))
     (flycheck-ert-should-syntax-check
      "language/javascript/warnings.js" flycheck-test-javascript-modes
-     '(3 2 warning "Use the function form of \"use strict\"." :id "strict"
+     '(3 2 warning "Use the function form of 'use strict'." :id "strict"
          :checker javascript-eslint)
-     '(4 9 warning "\"foo\" is defined but never used" :id "no-unused-vars"
+     '(4 9 warning "'foo' is defined but never used." :id "no-unused-vars"
          :checker javascript-eslint))))
```
Comment (lunaryorn) [inline test/flycheck-test.el]: These test changes appear to be related to eslint as well, aren't they?  Please move them to a separate pull request, too.

**Comment by lunaryorn** [comment]
@Simplify There are still some eslint-related changes, and I've suggested a couple of changes to the docstring—just minor improvements and markup fixes.

Please address these, otherwise LGTM and is definitely ready to go :)

**Comment by lunaryorn** [comment]
@cpitclaudel Mind to take another look?

**Comment by cpitclaudel** [comment]
LGTM.  This needs to be rebased, right? Good job in any case :)

**Comment by lunaryorn** [comment]
Oh, yes, my fix for the flawed order seems to have broken the merge :cry:

I'm sorry @Simplify but could you please rebase onto `master`?

**Comment by Simplify** [comment]
Ok, rebased and fixed.

**Comment by lunaryorn** [comment]
@Simplify Fine with me, feel free to merge :blush:

**Comment by lunaryorn** [comment]
@Simplify Thank you very much.

**Comment by Simplify** [comment]
Not a problem :) 
I'll submit ESLint fix PR today.
